plugin-azure-server/src/main/kotlin/jetbrains/buildServer/clouds/azure/arm/KubePodNameGenerator.kt

```diff
@@ -0,0 +1,134 @@
+                           @NotNull  executorServices: ExecutorServices) {
+
+    private val myCounters: ConcurrentMap<String, AtomicInteger> = ConcurrentHashMap()
+    private val myIndexTouchedMaps: ConcurrentMap<String, AtomicBoolean> = ConcurrentHashMap()
+    private var myIndexStorage: File = File(ServerPaths.PLUGIN_DATA_DIRECTORY_NAME, "azureKubeIndex")
+    private val myIsAvailable: AtomicBoolean = AtomicBoolean(false)
+    private val myLock: ReadWriteLock = ReentrantReadWriteLock()
+
+    init{
+        if (!myIndexStorage.exists()) {
+            myIndexStorage.mkdirs()
+        }
+        if (!myIndexStorage.isDirectory) {
+            throw CloudException("Unable to create a directory for azure arm plugin to monitor VM indexes")
+        }
+        myIsAvailable.set(true)
+
+        loadIndexes(myIndexStorage)
+
+        eventDispatcher.addListener(object : BuildServerAdapter() {
+            override fun serverShutdown() {
+                myIsAvailable.set(false)
+                storeIndexes()
+            }
+        })
+        executorServices.normalExecutorService.scheduleWithFixedDelay(this::storeIndexes, 60, 60, TimeUnit.SECONDS)
+    }
+
+    @Synchronized
+    private fun loadIndexes(@NotNull indexStorage: File) {
+        val indexes = indexStorage.listFiles() ?: return
+        for (indexFile in indexes) {
+            if (!indexFile.name.endsWith(".idx")) continue
+            val indexName = indexFile.name.substring(0, indexFile.name.length - 4)
+            try {
+                val indexValueFromFile = FileUtil.readText(indexFile).toInt()
+                if (indexValueFromFile > 0) {
+                    myCounters.putIfAbsent(indexName, AtomicInteger(0))
+                    myCounters[indexName]?.set(indexValueFromFile)
+                }
+            } catch (e: IOException) {
+                continue
+            }
+        }
+    }
+
+    private fun storeIndexes() {
+        // wait for generation operations to finish:
+        try {
+            if (!myLock.writeLock().tryLock(100, TimeUnit.MILLISECONDS)) {
+                Loggers.AGENT.warn("Waited more than 100ms to store Kube indexes")
+            }
+            for ((key, value) in myIndexTouchedMaps) {
+                if (value.compareAndSet(true, false)) {
+                    val counter = myCounters[key]
+                    try {
+                        val indexFile = File(myIndexStorage, "$key.idx")
+                        FileUtil.writeFile(indexFile,counter.toString(), Charset.defaultCharset().name())
+                    } catch (ignored: IOException) {
+                        continue
+                    }
+                }
+            }
+        } catch (e: InterruptedException) {
+            e.printStackTrace()
+        } finally {
+            myLock.writeLock().unlock()
+        }
+    }
+
+    fun generateNewVmName(image: AzureCloudImage): String {
```

Comment:
Since this class has concurrency it may be worth trying to push the concurrent work into a method that focuses on that. Right now this method sort of deals with concurrency and also deals with working out the name from the `image` and deals with trying to find if that image name has been used before. The problem with this is it makes it hard to understand the locking.

I think re-shuffling the code as below makes it a little easier to follow what the lock might be guarding.
```
fun getNextCounter(string name) {
  myLock.readLock().lock()
  try {
    verifyIsAvailable();
    int nextNumber = getNextCounterNoLock(name);
    markCounterAsTouched(name);
    return nextNumber();
    
  } finally {
      myLock.readLock().unlock();
  }
}
fun verifyIsAvailable() {
  if (!myIsAvailable.get()) {
                throw CloudException(
                    "Unable to generate a name for image " + image.id + " - server is shutting down"
                )
            }
}

private fun getNextCounterNoLock(prefix: String): Int {
        myCounters.putIfAbsent(prefix, AtomicInteger(0))
        return myCounters[prefix]!!.getAndIncrement()
    }
    
    private fun markCounterAsTouched(prefix: String) {
        myIndexTouchedMaps.putIfAbsent(prefix, AtomicBoolean(true))
        myIndexTouchedMaps[prefix]?.set(true)
    }
```

now `generateNewVmName` can just call `getNextCounter` without needing to worry about locks.

As a bonus, since the locking occurs closer to what it needs to protect we don't hold the lock while checking: `image.findInstanceById(newVmName)`.

Reply:
You are missing the cases where the lock is there when reading and storing data, which is added to the complexity. I like the verifyIsAvailable method and will look to pull that out.

Reply:
This snippet is just for shuffling the stuff in `generateNewVmName`, not for taking the locks away from `storeIndexes` and `loadIndexes` methods. Better keep those locks :D

Reply:
ah! but you are missing the while loop 
``` do {
                var prefix: String? = image.imageDetails.vmNamePrefix
                if (StringUtil.isEmptyOrSpaces(prefix)) {
                    prefix = image.imageDetails.sourceId
                }
                if (StringUtil.isEmptyOrSpaces(prefix) || prefix == null) {
                    return UUID.randomUUID().toString()
                }
                prefix = StringUtil.replaceNonAlphaNumericChars(prefix.trim().toLowerCase(), '-')
                newVmName = String.format("%s-%d", prefix, getNextCounter(prefix))
                setTouched(prefix)
            }
            while (image.findInstanceById(newVmName) != null)
```
which is definitely required.

Reply:
sorry I also didn't include the generateNewVmName() so in total it ends up looking like:
```
```
class AzureVmNameGenerator(@NotNull eventDispatcher: EventDispatcher<BuildServerListener>,
                           @NotNull  executorServices: ExecutorServices) {
... All the stuff above generateNewVmName() remains...

// This method is now all about working out a a vm name
// this method no longer cares about locking
// it calls getNextCounter() which deals with that locking stuff.
fun generateNewVmName(image: AzureCloudImage): String {
	var newVmName: String
	do {
		var prefix: String? = image.imageDetails.vmNamePrefix
		if (StringUtil.isEmptyOrSpaces(prefix)) {
			prefix = image.imageDetails.sourceId
		}
		if (StringUtil.isEmptyOrSpaces(prefix) || prefix == null) {
			return UUID.randomUUID().toString()
		}
		prefix = StringUtil.replaceNonAlphaNumericChars(prefix.trim().toLowerCase(), '-')
		newVmName = String.format("%s-%d", prefix, getNextCounter(prefix))
	}
	while (image.findInstanceById(newVmName) != null)
	return newVmName
}

// This method takes care of taking a lock and calling touch etc.
fun getNextCounter(string name) {
  myLock.readLock().lock()
  try {
    verifyIsAvailable();
    int nextNumber = getNextCounterNoLock(name);
    markCounterAsTouched(name);
    return nextNumber();
    
  } finally {
      myLock.readLock().unlock();
  }
}
fun verifyIsAvailable() {
  if (!myIsAvailable.get()) {
		throw CloudException(
			"Unable to generate a name for image " + image.id + " - server is shutting down"
		)
	}
}

private fun getNextCounterNoLock(prefix: String): Int {
	myCounters.putIfAbsent(prefix, AtomicInteger(0))
	return myCounters[prefix]!!.getAndIncrement()
}

private fun markCounterAsTouched(prefix: String) {
	myIndexTouchedMaps.putIfAbsent(prefix, AtomicBoolean(true))
	myIndexTouchedMaps[prefix]?.set(true)
}
```

I think that is a little easier to read.

Reply:
TBH I don't really like the getNextCounterNoLock and getNextCounter as two seperate methods. Thanks for the suggestion and I do like the look of generateNewVmName but I think I'll just leave it where it is!